### PR TITLE
Convert LandingPage to load data with a saga

### PR DESCRIPTION
**Pull request by kumar303**
Fixes https://github.com/mozilla/addons-frontend/issues/2859

The page looks the same as before except now there are loading placeholders and you can also see errors:

screenshot 2017-07-31 12 12 11
screenshot 2017-07-31 12 12 34

### CODE REVIEW

File: src/amo/actions/landing.js [hunk 1]
```diff
@@ -1,31 +1,38 @@
...
+    payload: { addonType, errorHandlerId },
   };
 }
 
 export function loadLanding({ addonType, featured, highlyRated, popular }) {
+  if (!addonType) {
+    throw new Error('addonType cannot be empty');
+  }
+  if (!featured) {
+    throw new Error('featured cannot be empty');
```
Comment (tofumatt) [inline src/amo/actions/landing.js]: Do you think adding `Addons` to the end of each variable would help with readability? `featured cannot be empty` is a bit ambiguous to me. 🤷‍♀️

File: src/amo/components/LandingAddonsCard/index.js [hunk 2]
```diff
@@ -13,10 +13,17 @@ export default class LandingAddonsCard extends React.Component {
     footerLink: PropTypes.object.isRequired,
     footerText: PropTypes.string.isRequired,
     header: PropTypes.node.isRequired,
+    loading: PropTypes.boolean,
+  }
+
+  static defaultProps = {
+    loading: false,
```
Comment (tofumatt) [inline src/amo/components/LandingAddonsCard/index.js]: Why is this needed? Don't we usually default to loading being false in the reducer?

File: src/amo/components/LandingPage/index.js [hunk 3]
```diff
@@ -5,22 +5,22 @@ import PropTypes from 'prop-types';
...
   SEARCH_SORT_TOP_RATED,
 } from 'core/constants';
-import { AddonTypeNotFound } from 'core/errors';
+import { withErrorHandler } from 'core/errorHandler';
 import log from 'core/logger';
 import {
-  apiAddonType as getApiAddonType,
-  safeAsyncConnect,
+  apiAddonType,
+  apiAddonTypeIsValid,
```
Comment (tofumatt) [inline src/amo/components/LandingPage/index.js]: Cannot wait for react-router 4 😆

File: src/amo/components/LandingPage/index.js [hunk 4]
```diff
@@ -37,61 +37,88 @@ const ICON_MAP = {
 
 export class LandingPageBase extends React.Component {
   static propTypes = {
-    apiAddonType: PropTypes.func.isRequired,
-    contentForType: PropTypes.func,
+    addonTypeOfResults: PropTypes.string,
```
Comment (tofumatt) [inline src/amo/components/LandingPage/index.js]: I would think `addonType` is sufficient because we use that everywhere else.

File: src/amo/components/LandingPage/index.js [hunk 5]
```diff
@@ -37,61 +37,88 @@ const ICON_MAP = {
...
+    errorHandler: PropTypes.object.isRequired,
     featuredAddons: PropTypes.array,
     highlyRatedAddons: PropTypes.array,
+    loading: PropTypes.boolean,
     popularAddons: PropTypes.array,
     i18n: PropTypes.object.isRequired,
     params: PropTypes.objectOf({
       visibleAddonType: PropTypes.string.isRequired,
     }).isRequired,
+    resultsLoaded: PropTypes.bool.isRequired,
```
Comment (tofumatt) [inline src/amo/components/LandingPage/index.js]: This could probably be computed from `featuredAddons.length || highlyRatedAddons.length || popularAddons.length` (with a helper function I guess?) rather than needing its own prop.

File: src/amo/components/LandingPage/index.js [hunk 6]
```diff
@@ -214,20 +233,23 @@ export class LandingPageBase extends React.Component {
           footerText={html.featuredFooterText}
           footerLink={html.featuredFooterLink}
           header={html.featuredHeader}
+          loading={loading}
```
Comment (tofumatt) [inline src/amo/components/LandingPage/index.js]: I think it's best to do in a future patch but technically these loading states shouldn't be shared; there are three API requests so we could load the data as it came in. I guess we'd want to A/B test that experience but my first thought is we should display data as we load it; if the featured data for some reason takes an extra `400ms` to load we should still display the other stuff, so arguably each request should have its own loading property.

File: src/amo/components/LandingPage/index.js [hunk 7]
```diff
@@ -37,61 +37,88 @@ const ICON_MAP = {
...
-    try {
-      const addonType = apiAddonType(params.visibleAddonType);
-      dispatch(setViewContext(addonType));
-    } catch (err) {
-      if (err instanceof AddonTypeNotFound) {
-        log.info('AddonTypeNotFound in setViewContextType()', err);
-      } else {
-        throw err;
-      }
+  getLandingDataIfNeeded() {
```
Comment (tofumatt) [inline src/amo/components/LandingPage/index.js]: Actually, if this function just took a `force = true` argument that made it load the data (which only `componentWillMount` would need here), you wouldn't have to deal with `resultsLoaded` at all. If the component mounts we always need to fetch data, and we only need to fetch new data if its updated and the `addonType` has changed.

Not sure if that's better. I think it's less code/state to manage but let me know what you think.

File: src/amo/sagas/landing.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      addonType ${addonType}: ${error}`);
+
+    yield put(errorHandler.createErrorAction(error));
+  } finally {
+    yield put(hideLoading());
+  }
+}
+
+export default function* landingSaga() {
+  yield takeEvery(LANDING_GET, fetchLandingAddons);
```
Comment (tofumatt) [inline src/amo/sagas/landing.js]: Arguably this could be `take` instead of `takeEvery` which would halt the old request if a user changed from one page to another. But this is probably fine–I think we do `takeEvery` everywhere and I don't think it'll be an issue...

File: tests/unit/amo/components/TestLandingPage.js [hunk 9]
```diff
@@ -17,31 +13,43 @@ import {
...
+      params: { visibleAddonType: visibleAddonType(ADDON_TYPE_EXTENSION) },
+      resultsLoaded: false,
+      ...props,
+    };
+  }
 
-    return findRenderedComponentWithType(renderIntoDocument(
+  function render(props = {}) {
+    return shallow(<LandingPageBase {...renderProps(props)} />, {
+      lifecycleExperimental: true,
```
Comment (tofumatt) [inline tests/unit/amo/components/TestLandingPage.js]: Hah, this explains why you could use `componentDidUpdate` and have them work with the tests. Neato. 👍

File: tests/unit/amo/helpers.js [hunk 10]
```diff
@@ -120,7 +120,8 @@ export function dispatchSearchResults({
   return { store };
 }
 
-export function createFeaturedApiResult(results) {
-  // Return a normalized result just like core.api.featured() does.
+export function createAddonsApiResult(results) {
```
Comment (tofumatt) [inline tests/unit/amo/helpers.js]: 👍

File: tests/unit/amo/reducers/testLanding.js [hunk 11]
```diff
@@ -96,7 +93,7 @@ describe('landing reducer', () => {
         },
       };
       const { highlyRated } = landing({
-        ...initialData,
+        ...initialState,
```
Comment (tofumatt) [inline tests/unit/amo/reducers/testLanding.js]: 👍

File: tests/unit/amo/sagas/testLanding.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+      apiState = state.api;
+      sagaTester = new SagaTester({
+        initialState: state,
+        reducers: { landing: landingReducer, api: apiReducer },
+      });
+
+      sagaTester.start(landingSaga);
+    });
+
+    function _getLanding(overrides = {}) {
```
Comment (tofumatt) [inline tests/unit/amo/sagas/testLanding.js]: This seems to be a common pattern in our tests, wonder if we could write a helper for it... or if it's not needed.

File: src/amo/components/LandingPage/index.js (hunk 5 above)
Comment (kumar303) [inline src/amo/components/LandingPage/index.js]: It cannot because the initial state of the reducer sets all of those to a zero-length array (which does not indicate that the request to load them has been fulfilled). Plus, if we add a new add-on result set then the if statement would need updating which seems clumsy to me.

File: src/amo/components/LandingPage/index.js (hunk 7 above)
Comment (kumar303) [inline src/amo/components/LandingPage/index.js]: I don't think that would handle the case where data was loaded on the server. In that case, `componentWillMount` should do nothing because all of the data has already been loaded.

File: src/amo/components/LandingPage/index.js (hunk 6 above)
Comment (kumar303) [inline src/amo/components/LandingPage/index.js]: Sure, that could be done as an optimization in the future but both the reducer and action creator would need refactoring because right now they 1) retrieve data from the 3 API calls at once and 2) do not store a distinct loading flag for each API result. It will also make the component more complex since it will have to juggle multiple loading flags. 

The important part is that the API requests are made in parallel so at least we have that so far.

File: tests/unit/amo/components/TestLandingPage.js (hunk 9 above)
Comment (kumar303) [inline tests/unit/amo/components/TestLandingPage.js]: Yeah, unfortunately it's not documented and the patch to do so is bitrotting :( https://github.com/airbnb/enzyme/pull/1033

File: tests/unit/amo/sagas/testLanding.js (hunk 12 above)
Comment (kumar303) [inline tests/unit/amo/sagas/testLanding.js]: I think we can do a pass soon on all the existing saga tests and come up with some shared helpers. I see a few patterns emerging.

File: src/amo/components/LandingPage/index.js (hunk 4 above)
Comment (kumar303) [inline src/amo/components/LandingPage/index.js]: I wanted to link this specifically to results because there is already `params.visibleAddonType` which is the `addonType` of the URL (i.e. what was requested). It's a bit confusing so I think it helps to tell them apart.

File: src/amo/components/LandingPage/index.js (hunk 5 above)
Comment (kumar303) [inline src/amo/components/LandingPage/index.js]: It might make things easier if we had one component per type of add-on result set (like `FeaturedAddons`, `HighlyRatedAddons`, and `PopularAddons`) then we could give them all their own state/reducer/saga as well and a lot of things would fall into place more naturally. I didn't want to change too much in this patch though.

File: src/amo/sagas/landing.js (hunk 8 above)
Comment (kumar303) [inline src/amo/sagas/landing.js]: Hmm. I'm not sure that `take` is right because it could make the page appear unresponsive for slow requests. It looks like we don't need to handle parallel requests though (and that could lead to confusing results) so I changed it to `takeLatest`. I think that will make the page still feel responsive if there is one hiccup request yet an otherwise snappy API. We'll probably have to tweak these later on after real world usage.

File: src/amo/components/LandingAddonsCard/index.js (hunk 2 above)
Comment (kumar303) [inline src/amo/components/LandingAddonsCard/index.js]: Ah, right, I suppose neither of those defaults are needed.

File: src/amo/components/LandingPage/index.js (hunk 4 above)
Comment (tofumatt) [inline src/amo/components/LandingPage/index.js]: Ah, okay.

File: src/amo/components/LandingPage/index.js (hunk 7 above)
Comment (tofumatt) [inline src/amo/components/LandingPage/index.js]: Ahhh, right. I didn't think of that.

**Comment by kumar303** [comment]
Thanks for the review!
